File: ports/java/src/core/graphics/BcsImage.java
package core.graphics;

import java.util.concurrent.locks.ReentrantLock;
import core.kernel.BcsObject;

public class BcsImage extends BcsObject {
    private final int width;
    private final int height;
    private final byte[] pixels;
    private final ReentrantLock lock = new ReentrantLock();

    public BcsImage(BcsObject parent, int width, int height) {
        super(parent);
        this.width = width;
        this.height = height;
        this.pixels = new byte[width * height * 4]; // RGBA
    }

    public int[] getSize() {
        lock.lock();
        try {
            return new int[]{width, height};
        } finally {
            lock.unlock();
        }
    }

    public void setPixel(int x, int y, byte r, byte g, byte b, byte a) {
        lock.lock();
        try {
            if (x >= 0 && x < width && y >= 0 && y < height) {
                int idx = (y * width + x) * 4;
                pixels[idx] = r;
                pixels[idx + 1] = g;
                pixels[idx + 2] = b;
                pixels[idx + 3] = a;
            }
        } finally {
            lock.unlock();
        }
    }

    public byte[] getPixels() {
        lock.lock();
        try {
            byte[] copy = new byte[pixels.length];
            System.arraycopy(pixels, 0, copy, 0, pixels.length);
            return copy;
        } finally {
            lock.unlock();
        }
    }
}

File: ports/csharp/src/core/graphics/BcsImage.cs
using System;
using Bcs.Core.Kernel;

namespace Bcs.Core.Graphics
{
    public class BcsImage : BcsObject
    {
        private int _width;
        private int _height;
        private byte[] _pixels;
        private readonly object _lock = new object();

        public BcsImage(BcsObject parent, int width, int height) : base(parent)
        {
            _width = width;
            _height = height;
            _pixels = new byte[width * height * 4]; // RGBA
        }

        public (int, int) Size()
        {
            lock (_lock)
            {
                return (_width, _height);
            }
        }

        public void SetPixel(int x, int y, byte r, byte g, byte b, byte a)
        {
            lock (_lock)
            {
                if (x >= 0 && x < _width && y >= 0 && y < _height)
                {
                    int idx = (y * _width + x) * 4;
                    _pixels[idx] = r;
                    _pixels[idx + 1] = g;
                    _pixels[idx + 2] = b;
                    _pixels[idx + 3] = a;
                }
            }
        }

        public byte[] Pixels()
        {
            lock (_lock)
            {
                byte[] copy = new byte[_pixels.Length];
                Array.Copy(_pixels, copy, _pixels.Length);
                return copy;
            }
        }
    }
}
